client/me/purchases/payment-methods/existing-credit-card.tsx

```diff
@@ -0,0 +1,397 @@
+	storedDetailsId: string;
+	paymentMethodToken: string;
+	paymentPartnerProcessorId: string;
+	activePayButtonText: string | undefined;
+} ): PaymentMethod {
+	debug( 'creating a new existing credit card payment method', {
+		id,
+		cardholderName,
+		cardExpiry,
+		brand,
+		last4,
+	} );
+
+	return {
+		id,
+		label: (
+			<ExistingCardLabel
+				last4={ last4 }
+				storedDetailsId={ storedDetailsId }
+				cardExpiry={ cardExpiry }
+				cardholderName={ cardholderName }
+				brand={ brand }
+				paymentPartnerProcessorId={ paymentPartnerProcessorId }
+			/>
+		),
+		submitButton: (
+			<ExistingCardPayButton
+				cardholderName={ cardholderName }
+				storedDetailsId={ storedDetailsId }
+				paymentMethodToken={ paymentMethodToken }
+				paymentPartnerProcessorId={ paymentPartnerProcessorId }
+				activeButtonText={ activePayButtonText }
+			/>
+		),
+		inactiveContent: (
+			<ExistingCardSummary
+				cardholderName={ cardholderName }
+				cardExpiry={ cardExpiry }
+				brand={ brand }
+				last4={ last4 }
+			/>
+		),
+		getAriaLabel: () => `${ brand } ${ last4 } ${ cardholderName }`,
+	};
+}
+
+function formatDate( cardExpiry: string ): string {
+	const expiryDate = new Date( cardExpiry );
+	const formattedDate = expiryDate.toLocaleDateString( 'en-US', {
+		month: '2-digit',
+		year: '2-digit',
+	} );
+
+	return formattedDate;
+}
+
+const CardDetails = styled.span`
+	display: inline-block;
+	margin-right: 8px;
+
+	.rtl & {
+		margin-right: 0;
+		margin-left: 8px;
+	}
+`;
+
+const CardHolderName = styled.span`
+	display: block;
+`;
+
+function ExistingCardLabel( {
+	last4,
+	cardExpiry,
+	cardholderName,
+	brand,
+	storedDetailsId,
+	paymentPartnerProcessorId,
+}: {
+	last4: string;
+	cardExpiry: string;
+	cardholderName: string;
+	brand: string;
+	storedDetailsId: string;
+	paymentPartnerProcessorId: string;
+} ): JSX.Element {
+	const { __, _x } = useI18n();
+
+	const [ isDialogVisible, setIsDialogVisible ] = useState( false );
+	const closeDialog = useCallback( () => setIsDialogVisible( false ), [] );
+	const queryClient = useQueryClient();
+
+	const queryKey = [ 'tax-info-is-set', storedDetailsId ];
+
+	const { data } = useQuery<
+		{ tax_postal_code: string; tax_country_code: string; is_tax_info_set: boolean },
+		Error
+	>( queryKey, () => fetchTaxInfo( storedDetailsId ), {} );
+
+	const mutation = useMutation(
+		( mutationInputValues: { tax_postal_code: string; tax_country_code: string } ) =>
+			setTaxInfo(
+				storedDetailsId,
+				mutationInputValues.tax_postal_code,
+				mutationInputValues.tax_country_code
+			),
+		{
+			onMutate: ( onMutateInputValues: { tax_postal_code: string; tax_country_code: string } ) => {
+				// Optimistically update the toggle
+				queryClient.setQueryData( queryKey, {
+					tax_postal_code: onMutateInputValues.tax_postal_code,
+					tax_country_code: onMutateInputValues.tax_country_code,
+					is_tax_info_set: true,
+				} );
+				closeDialog();
+			},
+			onSuccess: ( onSuccessInputValues: {
```

Comment:
I think we might need an `onError` handler too. I tried entering an invalid postal code and while I saw an error in the JS console, in the UI it appeared to work (although thankfully I wasn't able to assign the card with invalid information).

The error handler should be able to undo the optimistic update and display the error message from the server to the user.

Reply:
This should be working now. I've added the `<Notice>` to the dialog component, as it was causing some odd issues where it originally was.